### ISSUE TITLE
Bug: User.pk is assumed somewhat dangerously and many custom/3rd party fields won't work with allauth

### ISSUE THREAD

**Issue by elnygren** [pennersr/django-allauth#3419]
Both of the User pk<->str functions (utils.py) make quite strict assumptions about what the PK field is and tend to fail silently if it isn't what is expected.

```python
# utils.py

def user_pk_to_url_str(user):
    """
    This should return a string.
    """
    User = get_user_model()
    pk_field_class = type(User._meta.pk)
    if issubclass(pk_field_class, models.UUIDField):
        if isinstance(user.pk, str):
            return user.pk
        return user.pk.hex
    elif issubclass(pk_field_class, models.IntegerField):
        return int_to_base36(int(user.pk))
    return str(user.pk)


def url_str_to_user_pk(pk_str):
    User = get_user_model()
    remote_field = getattr(User._meta.pk, "remote_field", None)
    if remote_field and getattr(remote_field, "to", None):
        pk_field = User._meta.pk.remote_field.to._meta.pk
    else:
        pk_field = User._meta.pk
    pk_field_class = type(pk_field)
    if issubclass(pk_field_class, models.IntegerField):
        pk = base36_to_int(pk_str)
        # always call to_python() -- because there are fields like HashidField
        # that derive from IntegerField.
        pk = pk_field.to_python(pk)
    else:
        pk = pk_field.to_python(pk_str)
    return pk
```

For example `django-ulid` -> ULIDfield will not work and will cause subtle bugs. There are also other issues reported due to these:
* https://github.com/pennersr/django-allauth/issues/2004
* https://github.com/pennersr/django-allauth/issues/2156

**Comment by pennersr** [comment]
Can you please elaborate a bit and clearly detail what "will not work" and "cause subtle" means?

**Comment by pennersr** [comment]
Closing -- due to lack of input, assuming the fix for #3446 addresses this issue.

**Comment by pennersr** [comment]
https://codeberg.org/allauth/django-allauth/issues/3419
